src/main.rs

```diff
@@ -109,6 +120,158 @@ fn get_uuid(agent_uuid_config: &str) -> String {
+
+    Ok((unzipped, dec_payload_path, key_path))
+}
+
+// write symm key data and decrypted payload data out to specified files
+pub(crate) fn write_out_key_and_payload(
+    dec_payload: &[u8],
+    dec_payload_path: &str,
+    key: Arc<Mutex<[u8; KEY_LEN]>>,
+    key_path: &str,
+) -> Result<()> {
+    let symm_key = key.lock().unwrap(); //#[allow_ci]
+
+    let mut key_file = fs::File::create(key_path)?;
+    let bytes = key_file.write(&symm_key[..])?;
+    if bytes != symm_key.len() {
+        return Err(Error::Other(format!("Error writing symm key to {:?}: key len is {}, but {} bytes were written", key_path, symm_key.len(), bytes)));
+    }
+    info!("Wrote payload decryption key to {:?}", key_path);
+
+    let mut dec_payload_file = fs::File::create(dec_payload_path)?;
+    let bytes = dec_payload_file.write(dec_payload)?;
+    if bytes != dec_payload.len() {
+        return Err(Error::Other(format!("Error writing decrypted payload to {:?}: payload len is {}, but {} bytes were written", dec_payload_path, dec_payload.len(), bytes)));
+    }
+    info!("Wrote decrypted payload to {:?}", dec_payload_path);
+
+    Ok(())
+}
+
+// run a script (such as the init script, if any) and check the status
+pub(crate) fn run(dir: &str, script: &str, agent_uuid: &str) -> Result<()> {
+    let script_location = format!("{}/{}", dir, script);
+    info!("Running script: {:?}", script_location);
+
+    let script_path = Path::new(&script_location);
+    if !script_path.exists() {
+        return Err(Error::Other(format!("{} not found", script_location)));
+    }
+
+    if let Err(e) = Command::new("chmod")
+        .current_dir(dir)
+        .arg("u+x")
+        .arg(&script_location)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .status()
+    {
+        return Err(Error::Other(format!(
+            "unable to set {} as executable",
+            script_location
+        )));
+    }
+
+    match Command::new(&script_location)
+        .current_dir(dir)
+        .stdin(Stdio::piped())
+        .stdout(Stdio::piped())
+        .stderr(Stdio::piped())
+        .status()
+    {
+        Ok(_) => {
+            info!("{:?} ran successfully", script_location);
+            Ok(())
+        }
+        Err(e) => Err(Error::Other(format!(
+            "{:?} failed during run: {}",
+            script_location, e
+        ))),
+    }
+}
+
+// checks if keylime.conf indicates the payload should be unzipped, and does so if needed.
+// the input string is the directory where the unzipped file(s) should be stored.
+pub(crate) fn optional_unzip_payload(unzipped: &str) -> Result<()> {
+    let is_zip = config_get("cloud_agent", "extract_payload_zip")?;
+    if bool::from_str(&is_zip.to_lowercase())? {
+        let zipped_payload = config_get("cloud_agent", "dec_payload_file")?;
+        info!("Unzipping payload {} to {}", zipped_payload, unzipped);
+
+        let mut status = Command::new("unzip")
```

Comment:
There is nothing wrong with this, though I do want to ask: Why use the system command instead of a library?

Reply:
@ashcrow I tried a library, but it was a bit more fiddly than I expected and this just worked. I suppose we could always update it on a second pass if we want to make some enhancements.

Reply:
The zip_rs is not nice I agree, check out compress tools, I find it a lot simpler https://crates.io/crates/compress-tools

Reply:
@lukehinds Sure will do, I can push a new version of that commit if I get that working.

Reply:
The library seems straightforward, but I keep getting 

```
 INFO  keylime_agent                  > Unzipping payload decrypted_payload.zip to /tmp/secure/unzipped
Error: Io(Os { code: 2, kind: NotFound, message: "No such file or directory" })
```
Despite
```
[keylime-ima@localhost rust-keylime]$ ls /tmp/secure/unzipped/
decrypted_payload.zip  derived_tci_key
[keylime-ima@localhost rust-keylime]$ file /tmp/secure/unzipped/decrypted_payload.zip 
/tmp/secure/unzipped/decrypted_payload.zip: Zip archive data, at least v2.0 to extract
[keylime-ima@localhost rust-keylime]$ ls -la /tmp/secure/unzipped/decrypted_payload.zip 
-rw-r--r--. 1 root root 8738 Jul 12 14:02 /tmp/secure/unzipped/decrypted_payload.zip
```

The code:
```
let zipped_payload = config_get("cloud_agent", "dec_payload_file")?;
        info!("Unzipping payload {} to {}", &zipped_payload, &unzipped);

        let mut source = fs::File::open(&zipped_payload)?;    
        assert!(Path::new(&zipped_payload).exists());

        let dest = Path::new(&unzipped);
        assert!(Path::new(&dest).exists());

        let file_list = list_archive_files(&mut source).expect("failed to list files");
        info!("File list: {:?}", file_list);

        uncompress_archive(&mut source, &dest, Ownership::Ignore).expect("failed to unzip");
```

Still debugging...

Reply:
Okay, it worked with a test example so it should be doable here shortly.

Reply:
It says it succeeds, but then the unzipped files aren't there. All I can think of is maybe it isn't opening with the correct permissions, as it's apparently `r` only. This should be easy to update but for some reason is not.

Reply:
I figured out the error... apparently `compress_tools::list_archive_files` mutates the `src` archive you give it so that there are no files left afterwards. Bizarre!